exporter/elasticexporter/internal/translator/elastic/traces.go

```diff
@@ -0,0 +1,467 @@
+	if name == "" {
+		return
+	}
+	c.serviceFramework.Name = truncate(name)
+	c.serviceFramework.Version = truncate(version)
+	c.service.Framework = &c.serviceFramework
+	c.model.Service = &c.service
+}
+
+func (c *transactionContext) setHTTPMethod(method string) {
+	c.request.Method = truncate(method)
+	c.model.Request = &c.request
+}
+
+func (c *transactionContext) setHTTPScheme(scheme string) {
+	c.request.URL.Protocol = truncate(scheme)
+	c.model.Request = &c.request
+}
+
+func (c *transactionContext) setHTTPURL(httpURL string) error {
+	u, err := url.Parse(httpURL)
+	if err != nil {
+		return err
+	}
+	// http.url is typically a relative URL, i.e. missing
+	// the scheme and host. Don't override those parts of
+	// the URL if they're empty, as they may be set by
+	// other attributes.
+	if u.Scheme != "" {
+		c.request.URL.Protocol = truncate(u.Scheme)
+	}
+	if hostname := u.Hostname(); hostname != "" {
+		c.request.URL.Hostname = truncate(hostname)
+	}
+	if port := u.Port(); port != "" {
+		c.request.URL.Port = truncate(u.Port())
+	}
+	c.request.URL.Path = truncate(u.Path)
+	c.request.URL.Search = truncate(u.RawQuery)
+	c.request.URL.Hash = truncate(u.Fragment)
+	c.model.Request = &c.request
+	return nil
+}
+
+func (c *transactionContext) setHTTPHost(hostport string) error {
+	host, port, err := net.SplitHostPort(hostport)
+	if err != nil {
+		return err
+	}
+	c.request.URL.Hostname = truncate(host)
+	c.request.URL.Port = truncate(port)
+	c.model.Request = &c.request
+	return nil
+}
+
+func (c *transactionContext) setHTTPHostname(hostname string) {
+	c.request.URL.Hostname = truncate(hostname)
+	c.model.Request = &c.request
+}
+
+func (c *transactionContext) setHTTPVersion(version string) {
+	c.request.HTTPVersion = truncate(version)
+	c.model.Request = &c.request
+}
+
+func (c *transactionContext) setHTTPRemoteAddr(remoteAddr string) {
+	c.requestSocket.RemoteAddress = truncate(remoteAddr)
+	c.request.Socket = &c.requestSocket
+	c.model.Request = &c.request
+}
+
+func (c *transactionContext) setHTTPRequestHeader(k string, v ...string) {
+	for i := range v {
+		v[i] = truncate(v[i])
+	}
+	c.request.Headers = append(c.request.Headers, model.Header{Key: truncate(k), Values: v})
+	c.model.Request = &c.request
+}
+
+func (c *transactionContext) setHTTPStatusCode(statusCode int) {
+	c.response.StatusCode = statusCode
+	c.model.Response = &c.response
+}
+
+type spanContext struct {
+	model   model.SpanContext
+	http    model.HTTPSpanContext
+	httpURL url.URL
+	db      model.DatabaseSpanContext
+}
+
+func (c *spanContext) modelContext() *model.SpanContext {
+	switch {
+	case c.model.HTTP != nil:
+	case c.model.Database != nil:
+	case len(c.model.Tags) != 0:
+	default:
+		return nil
+	}
+	return &c.model
+}
+
+func (c *spanContext) setHTTPStatusCode(statusCode int) {
+	c.http.StatusCode = statusCode
+	c.model.HTTP = &c.http
+}
+
+func (c *spanContext) setHTTPURL(httpURL string) error {
```

Comment:
I can't quite follow why the logic inside the `transaction` and `span` `setHTTPURL` methods differs e.g. `RawQuery` vs `Search`, ..).

Reply:
This is due to the fact that transactions and spans in our model have differing levels of detail for URLs. Transaction URLs are exploded into their parts, span URLs are not. The "search", "hash", etc. names are historical. We can tidy this up when we remove the dependency on the Go agent.